DOCSP-31806 ejson methods (#284)
mongodb/docs-mongodb-shell: source/reference/ejson/stringify.txt

.. _mongosh-ejson-stringify:

=================
EJSON.stringify()
=================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

The ``EJSON.stringify()`` method converts BSON values to strings. 

Syntax
------

The ``EJSON.stringify()`` method takes a BSON object as input and
optional modifiers that control the format of the output string.

.. code-block:: javascript
   :copyable: false

   EJSON.stringify(BSON object, [replacer], [space], [options])

Command Fields
--------------

The ``EJSON.stringify()`` method has these fields:

.. list-table::
   :header-rows: 1

   * - Field
     - Type
     - Necessity
     - Description

   * - ``value``
     -  BSON object
     -  Required
     -  The object ``EJSON.stringify`` transforms

   * - ``replacer``
     -  array or function
     -  Optional
     -  Modifies output. If the value exists but it is not an array or a

>         function, ``EJSON.stringify`` returns all document fields. 

Reviewer: [suggestion]

Consider updating to ``EJSON.stringify()`` to be more consistent.

Author: Done

           :widths: 25 25 50

           * - Option
             - Default
             - Meaning

           * - ``relaxed``
             - ``true``
             - Enable relaxed mode for Extended JSON. Returns native 
               JSON types instead of attaching BSON type information
               where applicable.

Behavior
--------

You can call the EJSON interface from inside an interactive ``mongosh``
session or from the system command line using ``--eval``.

Call the EJSON interface from an interactive session: 

.. code-block:: javascript

   EJSON.stringify( db.sales.find().toArray(), null, 2  )

Call the EJSON interface from the system command line: 

.. code-block:: shell

   mongosh --eval "EJSON.stringify( db.sales.find().toArray(), null, 2  )"

To control how documents are passed to EJSON, use one of the ``mongosh``
:ref:`cursor method <doc-cursor-methods>` iterators. 

.. list-table::
   :header-rows: 1

   * - Iterator
     - Characteristics

   * - :method:`.toArray() <cursor.toArray>`
     - Blocking, buffers the entire result

   * - :method:`.foreach <cursor.forEach()>`
     - Non-blocking, prints documents one-by-one

   * - :method:`.next() <cursor.next>`

>      -  Non-blocking, manually iterate over results

Reviewer: [spacing nit]

Extra space between bullet and "Non-blocking".

Author: Done

.. code-block:: javascript

   [
      {
         "custId": 345,
         "purchaseDate": {
            "$date": "2023-07-04T00:00:00Z"
         },
         "quantity": 8,
         "cost": {
            "$numberDecimal": "100.60"
         }
      },
      {
         "custId": 346,
         "purchaseDate": {
            "$date": "2023-07-12T00:00:00Z"
         },
         "quantity": 6,
         "cost": {
            "$numberDecimal": "175.45"
         }
      },
      {
         "custId": 486,
         "purchaseDate": {
            "$date": "2023-08-01T00:00:00Z"
         },
         "quantity": 18,
         "cost": {
            "$numberDecimal": "200.53"
         }
      }
   ]

The ``replacer`` function updates two fields, ``_id`` and ``quantity``.

``EJSON.stringify()`` ignores fields with undefined values, so setting
``_id: undefined`` removes the ``_id`` field from the output string.

The function also modifies the ``quantity`` field in the output string.
All of the ``quantity`` values are multiplied by two in the output
string. ``EJSON.stringify()`` does not update the collection.

Write to a File from Inside mongosh
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> To write to a file from within ``mongosh``, call the ``fs`` API. Use

Reviewer: [question]

Would it be too general to say "use the ``fs`` API here?"

Author: Done - I think they both work in this context

The output is: 

.. code-block:: json
   :copyable: false 

   # Note: This is only the first document.

   {
      "_id": {
         "$oid": "64da90c1175f5091debcab26"
      },
      "custId": {
         "$numberInt": "345"
      },
      "purchaseDate": {
         "$date": {
            "$numberLong": "1688428800000"
         }
      },
      "quantity": {
         "$numberInt": "4"
      },
      "cost": {
         "$numberDecimal": "100.60"
      }
   }

Filter Output Fields
~~~~~~~~~~~~~~~~~~~~

``EJSON.stringify()`` provides formatting options that reduce the need
for an additional JSON parser like ``jq``. 

.. code-block:: shell

   # Note: This example is formatted to fit on the page.

   mongosh --quiet \
           --eval "EJSON.stringify( \
                db.sales.find( {}, \
                   { _id: 0, custId: 1, quantity: 1 } ).toArray(), null, 2 \
             );"

The output is: 

.. code-block:: json
   :copyable: false 

   [
     {
       "custId": 345,
       "quantity": 4
     },
     {
       "custId": 346,
       "quantity": 3
     },
     {
       "custId": 486,
       "quantity": 9
     }
   ]

Legacy ``tojsononeline()``
~~~~~~~~~~~~~~~~~~~~~~~~~~

> The ``mongosh`` formats query output differently than the deprecated

Reviewer: [suggestion]

Consider removing "The" here.

Author: Done

``mongo`` shell. You may have scripts which rely on the old format. To
reformat ``mongosh`` output on a single line, use ``EJSON.stringify()``.

Run a sample query in ``mongosh`` and ``mongo`` to see the different
formats.

.. code-block:: javascript

   db.sales.find( { custId: 345 } )

Legacy output:

.. code-block:: javascript
   :copyable: false

   { "_id" : ObjectId("64da90c1175f5091debcab26"), "custId" : 345, "purchaseDate" : ISODate("2023-07-04T00:00:00Z"), "quantity" : 4, "cost" : NumberDecimal("100.60") }

``mongosh`` output:

.. code-block:: javascript
   :copyable: false

   db.sales.find( { custId: 345 } )
   [
     {
       _id: ObjectId("64da90c1175f5091debcab26"),
       custId: 345,
       purchaseDate: ISODate("2023-07-04T00:00:00.000Z"),
       quantity: 4,
       cost: Decimal128("100.60")
     }
   ]

> Reformat the output with the EJSON ``stringify()`` method.

Reviewer: [suggestion]

Consider expressing this as ``EJSON.stringify()`` to be more consistent with the surrounding copy.

Author: Done

             - Effect

           * - array
             - An array of document fields to include in the output

           * - function
             - A function that takes two parameters, ``key`` and
               ``value``. ``key`` provides the function's ``this``
               context. EJSON returns the transformed ``value``. See
               below for an :ref:`example <ex-ejson-function>`   

   * - ``spacer``
     -  integer or string
     -  Optional
     -  Controls spacing in the output. Use ``null`` as a place holder
        for ``replacer`` if you only want to specify the ``spacer``
        option.
 
        .. list-table::
           :header-rows: 1
           :widths: 30 70

           * - Value
             - Effect

           * - integer
             - The number of spaces to indent each level. ``10`` is the
               maximum.

           * - string
             - The character to use to indent each level.

   * - ``options``
     -  boolean
     -  Optional
     -  Additional configuration options
      
        .. list-table::
           :header-rows: 1
           :widths: 25 25 50

           * - Option
             - Default
             - Meaning

>            * - ``legacy``
>              - ``false``
>              - Use the Extended JSON v1 specification for output.
> 

Reviewer 2: I wouldn’t document this, there’s no good reason why mongosh users would use this option :slightly_smiling_face: 

```suggestion
```

           * - ``relaxed``

>              - ``false``
>              - Enable relaxed mode for Extended JSON. Returns native 
>                JavaScript types instead of BSON types when possible.

Reviewer 2: ```suggestion
             - ``true``
             - Enable relaxed mode for Extended JSON. Returns native 
               JSON types instead of attaching BSON type information where applicable.
```

> 
>            * - ``useBigint64`` 
>              - ``false``
>              - Enable native bigint support.

Reviewer 2: This is not a `stringify()` option:

```suggestion
```

Examples
--------

To try these examples, first create the ``sales`` collection:

.. code-block:: javascript

   db.sales.insertMany( [ 
      { custId: 345, purchaseDate: ISODate("2023-07-04"), quantity: 4, cost: Decimal128("100.60"), },
      { custId: 346, purchaseDate: ISODate("2023-07-12"), quantity: 3, cost: Decimal128("175.45"), },
      { custId: 486, purchaseDate: ISODate("2023-08-01"), quantity: 9, cost: Decimal128("200.53"), },
   ] )

Change Output Spacing
~~~~~~~~~~~~~~~~~~~~~

To increase indentation between levels, set the ``spacing`` option.

.. code-block:: javascript

   EJSON.stringify( db.sales.findOne( { custId: 345 } ), null , 5 )

``EJSON.stringify()`` indents each document level five spaces.

.. code-block:: javascript

   {
        "_id": {
             "$oid": "64da90c1175f5091debcab26"
        },
        "custId": 345,
        "purchaseDate": {
             "$date": "2023-07-04T00:00:00Z"
        },
        "quantity": 4,
        "cost": {
             "$numberDecimal": "100.60"
        }
   }

To highlight the indentation between levels, use the ``spacing`` option
to set a string of dots.

.. code-block:: javascript

>    EJSON.stringify( db.sales.findOne( { custId: 345 } ), null , "..." )

Reviewer 2: I’m not sure if this is a good example because the result isn’t valid JSON and cannot be parsed by anything.

A possible use case for using a string as the indentation argument could be passing `'\t'` in order to get tab-indented output, but really, I would consider just not documenting this (and linking to the [MDN docs for `JSON.stringify()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify) instead if we feel that this is important to have referenced somewhere).
